__torch_function__ mode

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75154

`__torch_function__` mode brings `enable_python_mode` functionality to the `__torch_function__` level of overrides (at the Python API level). I intend to use this to collect test data for validating abstract interpreters we plan to write for operators, but there are many other possible applications (for example, tracers like FX could use this to track factory function calls without monkey-patching the PyTorch namespace, and the mode could be used to vary backward semantics of functions without having to add new arguments to them, see https://github.com/albanD/subclass_zoo/blob/main/inner_autograd_tensor.py)

This PR also gives a design for mode stacks which is implemented in Python (rather in some sort of C++ infrastructure). My goal here was to trade off performance so that I could simplify the C++ bindings (we only expose `_set_torch_function_mode` and `_get_torch_function_mode` in the bindings) and iterate quickly on a more sophisticated mode stack design. The design here is oriented around TorchFunctionMode objects, which are chained together via object composition. IF this design is accepted, we should also apply it to the existing `__torch_dispatch__` mode.

**What is `__torch_function__` mode?** Check the documentation for TorchFunctionMode in torch/overrides.py

**How does it work?** There are two parts to the implementation, a simple but non-compositional C++ API, and a compositional Python API built on top of the C++ API.

In C++:

* We add a new Python mode object to PythonTorchFunctionTLS; the object doesn't have any type constraints but we will assume that it has a `__torch_function__` method defined on it. (Unlike `python_eager_mode`, this object need not be a class; it can be an honest to goodness instance of a class.)
* We modify `has_torch_function` to return true if the mode is non-null. In some cases, this is wrong (specifically when we are accumulating overloaded arguments), so those get a special argument to say "ignore modes when testing." I didn't do a very careful audit of all the sites so this is probably where we mostly likely have bugs in the implementation.
* When handling `__torch_function__`, we check if this mode is set. If it is, we first call `__torch_function__` on the mode to attempt to handle the operation (that is to say, we process modes first, and then concrete values)--unusually, we also reset the mode to be empty before calling back into `__torch_function__` (see https://github.com/pytorch/pytorch/issues/75166 for back story). If the mode returns NotImplemented we attempt to handle the operator with the subclasses instead. I fixed a bug in the error reporting where pointer values rather than actual reprs were printed.

In Python:

* We add a new class hierarchy TorchFunctionMode, representing objects that can handle `__torch_function__` and can chain with each other (allowing for nested modes). These classes have a metaclass TorchFunctionModeMeta which is (1) responsible for adding chaining functionality, without requiring users to manually record `inner` themselves, and (2) responsible for resetting the current mode (which is null at the time `__torch_function__` is called) to the mode of the *inner* mode class. We represent `__torch_function__` handlers as objects, not classes, so that a single handler can be associated with arbitrary state (most notably, the `inner` property; but you can define a parametrized modes--see example in test suite).
* We then provide two user-visible functions for working with the mode stack: `push_torch_function_mode` is the recommended function which is compositional (allowing for nested modes) as well as `enable_torch_function_mode` which is non-compositional (good for simple use cases, or swapping out a known mode with another mode). Check the docblocks in the PR for more details.
* For symmetry with `__torch_dispatch__` mode, we also accept a Tensor subclass as a mode passed to the argument; this works in much the same way a non-compositional TorchFunctionMode would be applied.

**Where there alternatives?**

* `push_torch_function_mode` could have been implemented as a C++ level concept, representing the stack explicitly in C++. This would have involved writing quite a lot of complex C++ binding code, and when I determined that I could implement all of this in Python using only two simple bindings I opted for this strategy for development speed. It may be useful to do it this way in the future for performance reasons.
* This PR introduces a new concept, TorchFunctionMode, whereas the existing `__torch_dispatch__` mode reused the class object for specifying the mode. I think introducing a new concept here is warranted for the following reasons: (1) `enable_python_mode` awkwardly requires you to subclass Tensor to write your mode, even if you never intend to instantiate your tensor, (2) there is no way to add extra metadata to a class without dynamically generating a class on the fly; this metadata is required for compositional use cases (aka, `inner` ala functorch), (3) even if you were able to dynamically generate the class working up and sideways the class hierarchy with super() and CurrentClass is awkward and people are unlikely to do it correctly. This is a case of favoring composition over inheritance, giving us an easier to understand API (even if it is not parsimonious).

**Examples?** Check the test cases in TestTorchFunctionMode in test/test_overrides.py

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
__torch_function__ mode

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "__torch_function__ mode"


Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "__torch_function__ mode"


Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "__torch_function__ mode"


Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "__torch_function__ mode"


Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "__torch_function__ mode"


Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "__torch_function__ mode"


Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "__torch_function__ mode"


Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "__torch_function__ mode"


Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
cc'ing @mruberry as you may find the API useful

ezyang committed:
Update on "__torch_function__ mode"


`__torch_function__` mode brings `enable_python_mode` functionality to the `__torch_function__` level of overrides (at the Python API level). I intend to use this to collect test data for validating abstract interpreters we plan to write for operators, but there are many other possible applications (for example, tracers like FX could use this to track factory function calls without monkey-patching the PyTorch namespace, and the mode could be used to vary backward semantics of functions without having to add new arguments to them, see https://github.com/albanD/subclass_zoo/blob/main/inner_autograd_tensor.py)

This PR also gives a design for mode stacks which is implemented in Python (rather in some sort of C++ infrastructure). My goal here was to trade off performance so that I could simplify the C++ bindings (we only expose `_set_torch_function_mode` and `_get_torch_function_mode` in the bindings) and iterate quickly on a more sophisticated mode stack design. The design here is oriented around TorchFunctionMode objects, which are chained together via object composition. IF this design is accepted, we should also apply it to the existing `__torch_dispatch__` mode.

**What is `__torch_function__` mode?** Check the documentation for TorchFunctionMode in torch/overrides.py

**How does it work?** There are two parts to the implementation, a simple but non-compositional C++ API, and a compositional Python API built on top of the C++ API.

In C++:

* We add a new Python mode object to PythonTorchFunctionTLS; the object doesn't have any type constraints but we will assume that it has a `__torch_function__` method defined on it. (Unlike `python_eager_mode`, this object need not be a class; it can be an honest to goodness instance of a class.)
* We modify `has_torch_function` to return true if the mode is non-null. In some cases, this is wrong (specifically when we are accumulating overloaded arguments), so those get a special argument to say "ignore modes when testing." I didn't do a very careful audit of all the sites so this is probably where we mostly likely have bugs in the implementation.
* When handling `__torch_function__`, we check if this mode is set. If it is, we first call `__torch_function__` on the mode to attempt to handle the operation (that is to say, we process modes first, and then concrete values)--unusually, we also reset the mode to be empty before calling back into `__torch_function__` (see https://github.com/pytorch/pytorch/issues/75166 for back story). If the mode returns NotImplemented we attempt to handle the operator with the subclasses instead. I fixed a bug in the error reporting where pointer values rather than actual reprs were printed.

In Python:

* We add a new class hierarchy TorchFunctionMode, representing objects that can handle `__torch_function__` and can chain with each other (allowing for nested modes). These classes have a metaclass TorchFunctionModeMeta which is (1) responsible for adding chaining functionality, without requiring users to manually record `inner` themselves, and (2) responsible for resetting the current mode (which is null at the time `__torch_function__` is called) to the mode of the *inner* mode class. We represent `__torch_function__` handlers as objects, not classes, so that a single handler can be associated with arbitrary state (most notably, the `inner` property; but you can define a parametrized modes--see example in test suite).
* We then provide two user-visible functions for working with the mode stack: `push_torch_function_mode` is the recommended function which is compositional (allowing for nested modes) as well as `enable_torch_function_mode` which is non-compositional (good for simple use cases, or swapping out a known mode with another mode). Check the docblocks in the PR for more details.
* For symmetry with `__torch_dispatch__` mode, we also accept a Tensor subclass as a mode passed to the argument; this works in much the same way a non-compositional TorchFunctionMode would be applied.

**Where there alternatives?**

* `push_torch_function_mode` could have been implemented as a C++ level concept, representing the stack explicitly in C++. This would have involved writing quite a lot of complex C++ binding code, and when I determined that I could implement all of this in Python using only two simple bindings I opted for this strategy for development speed. It may be useful to do it this way in the future for performance reasons.
* This PR introduces a new concept, TorchFunctionMode, whereas the existing `__torch_dispatch__` mode reused the class object for specifying the mode. I think introducing a new concept here is warranted for the following reasons: (1) `enable_python_mode` awkwardly requires you to subclass Tensor to write your mode, even if you never intend to instantiate your tensor, (2) there is no way to add extra metadata to a class without dynamically generating a class on the fly; this metadata is required for compositional use cases (aka, `inner` ala functorch), (3) even if you were able to dynamically generate the class working up and sideways the class hierarchy with super() and CurrentClass is awkward and people are unlikely to do it correctly. This is a case of favoring composition over inheritance, giving us an easier to understand API (even if it is not parsimonious).

**Examples?** Check the test cases in TestTorchFunctionMode in test/test_overrides.py

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "__torch_function__ mode"


`__torch_function__` mode brings `enable_python_mode` functionality to the `__torch_function__` level of overrides (at the Python API level). I intend to use this to collect test data for validating abstract interpreters we plan to write for operators, but there are many other possible applications (for example, tracers like FX could use this to track factory function calls without monkey-patching the PyTorch namespace, and the mode could be used to vary backward semantics of functions without having to add new arguments to them, see https://github.com/albanD/subclass_zoo/blob/main/inner_autograd_tensor.py)

This PR also gives a design for mode stacks which is implemented in Python (rather in some sort of C++ infrastructure). My goal here was to trade off performance so that I could simplify the C++ bindings (we only expose `_set_torch_function_mode` and `_get_torch_function_mode` in the bindings) and iterate quickly on a more sophisticated mode stack design. The design here is oriented around TorchFunctionMode objects, which are chained together via object composition. IF this design is accepted, we should also apply it to the existing `__torch_dispatch__` mode.

**What is `__torch_function__` mode?** Check the documentation for TorchFunctionMode in torch/overrides.py

**How does it work?** There are two parts to the implementation, a simple but non-compositional C++ API, and a compositional Python API built on top of the C++ API.

In C++:

* We add a new Python mode object to PythonTorchFunctionTLS; the object doesn't have any type constraints but we will assume that it has a `__torch_function__` method defined on it. (Unlike `python_eager_mode`, this object need not be a class; it can be an honest to goodness instance of a class.)
* We modify `has_torch_function` to return true if the mode is non-null. In some cases, this is wrong (specifically when we are accumulating overloaded arguments), so those get a special argument to say "ignore modes when testing." I didn't do a very careful audit of all the sites so this is probably where we mostly likely have bugs in the implementation.
* When handling `__torch_function__`, we check if this mode is set. If it is, we first call `__torch_function__` on the mode to attempt to handle the operation (that is to say, we process modes first, and then concrete values)--unusually, we also reset the mode to be empty before calling back into `__torch_function__` (see https://github.com/pytorch/pytorch/issues/75166 for back story). If the mode returns NotImplemented we attempt to handle the operator with the subclasses instead. I fixed a bug in the error reporting where pointer values rather than actual reprs were printed.

In Python:

* We add a new class hierarchy TorchFunctionMode, representing objects that can handle `__torch_function__` and can chain with each other (allowing for nested modes). These classes have a metaclass TorchFunctionModeMeta which is (1) responsible for adding chaining functionality, without requiring users to manually record `inner` themselves, and (2) responsible for resetting the current mode (which is null at the time `__torch_function__` is called) to the mode of the *inner* mode class. We represent `__torch_function__` handlers as objects, not classes, so that a single handler can be associated with arbitrary state (most notably, the `inner` property; but you can define a parametrized modes--see example in test suite).
* We then provide two user-visible functions for working with the mode stack: `push_torch_function_mode` is the recommended function which is compositional (allowing for nested modes) as well as `enable_torch_function_mode` which is non-compositional (good for simple use cases, or swapping out a known mode with another mode). Check the docblocks in the PR for more details.
* For symmetry with `__torch_dispatch__` mode, we also accept a Tensor subclass as a mode passed to the argument; this works in much the same way a non-compositional TorchFunctionMode would be applied.

**Where there alternatives?**

* `push_torch_function_mode` could have been implemented as a C++ level concept, representing the stack explicitly in C++. This would have involved writing quite a lot of complex C++ binding code, and when I determined that I could implement all of this in Python using only two simple bindings I opted for this strategy for development speed. It may be useful to do it this way in the future for performance reasons.
* This PR introduces a new concept, TorchFunctionMode, whereas the existing `__torch_dispatch__` mode reused the class object for specifying the mode. I think introducing a new concept here is warranted for the following reasons: (1) `enable_python_mode` awkwardly requires you to subclass Tensor to write your mode, even if you never intend to instantiate your tensor, (2) there is no way to add extra metadata to a class without dynamically generating a class on the fly; this metadata is required for compositional use cases (aka, `inner` ala functorch), (3) even if you were able to dynamically generate the class working up and sideways the class hierarchy with super() and CurrentClass is awkward and people are unlikely to do it correctly. This is a case of favoring composition over inheritance, giving us an easier to understand API (even if it is not parsimonious).

**Examples?** Check the test cases in TestTorchFunctionMode in test/test_overrides.py

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -206,23 +217,83 @@ static PyObject* get_type_of_overloaded_arg(PyObject* obj_or_type) {
+      // Disable mode on the inside; this makes for a more user-friendly
+      // experience if you try to, e.g., print your tensors.
+      torch::overrides::no_torch_function_mode g;
+      // Blegh.  This accidentally works in PyObject_CallFunctionObjArgs below
```
No I just want the dict kwarg to always be passed. Will also save people from having to default the kwargs argument which is a pain.

ezyang commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -206,23 +217,83 @@ static PyObject* get_type_of_overloaded_arg(PyObject* obj_or_type) {
+  const char* torch_function_name_str = nullptr;
+  switch (torch_function_name) {
+    case TorchFunctionName::TorchFunction:
+      torch_function_name_str = "__torch_function__";
```
Nope. torch_function_mode_name was when I was thinking about changing the function name for TorchDispatchMode objects. We're not allowed to change the name of the torch function on Tensor-like objects though. (Do you think TorchDispatchMode should define `__torch_function__`, or something differently named?)

ezyang replied:
> But we will never change it, so I guess it is not really important to have it defined only once.

Yeah, we're not going to change it. Hard code away.

However, we *do* have a choice before landing this PR, which is whether or not to have the mode objects define `__torch_function__`. What is weird about reusing the name in this way is I can also pass a mode object AS AN ARGUMENT to a torch API function, and it will trigger the alternate codepath and stick you in the `__torch_function__` which will promptly fail to do anything reasonable because it was expecting to be used as a mode, and not expecting to see a mode object as one of the arguments.

The reason I didn't change this name for the PR was that it seemed unlikely real users would make this mistake, and I really didn't want to have to explain *yet another* magic method name to users of this feature.

ezyang commented on torch/csrc/utils/python_torch_function_mode.h:
```diff
@@ -0,0 +1,16 @@
+namespace torch {
+namespace overrides {
+
+struct no_torch_function_mode {
```
C++ side is oblivious to the inner concept. To C++ it makes sense to clear the entire mode, that's all it knows about. Python-side is what has to work hard!

ezyang commented on torch/overrides.py:
```diff
@@ -1384,6 +1385,15 @@ def handle_torch_function(
+    mode = _get_torch_function_mode()
+    if mode is not None:
+        # NB: unlike on tensors, modes are instances
+        with _no_torch_function_mode():
```
It's all TLS, how could there be a race? (Do you want an assert?)

albanD replied:
Race was a bad word, it was more that these APIs needs to be done in a specific order. But since they are private and not exposed to the user that's fine.

I think I was also thrown off by the "stack only in python" idea.
Sounds good

ezyang commented on torch/overrides.py:
```diff
@@ -1587,3 +1601,227 @@ def is_tensor_like(inp):
+        (isinstance(mode, type) and issubclass(mode, torch.Tensor))
+    ):
+        raise TypeError(
+            "expected to get None, TorchFunctionMode or Tensor subclass as argument, got "
```
That's right. I didn't think this through carefully enough. I am afraid to accept generic class here as you could accidentally pass in type representing subclass of TorchFunctionMode and that would be an error. Maybe we can test that it is NOT that.

ezyang replied:
(The confusion here is also another reason why I was considering naming the `__torch_function__` something different on mode classes)

ezyang commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -235,15 +306,24 @@ auto handle_torch_function_no_python_arg_parser(at::ArrayRef<py::handle> overloa
+      // a bit safe, it doesn't cost us anything since this is error reporting
+      const auto& maybe_mode = at::impl::PythonTorchFunctionTLS::get_mode();
+      TORCH_INTERNAL_ASSERT(mode_obj == maybe_mode->ptr(getPyInterpreter()));
+      ss << " nor was it found on the currently active mode "
```
yes? though not in the current tests

ezyang commented on aten/src/ATen/PythonTorchFunctionTLS.h:
```diff
@@ -9,11 +10,16 @@ struct TORCH_API PythonTorchFunctionTLS {
 
 private:
   bool disabled_;
+  std::shared_ptr<c10::SafePyObject> mode_;
```
Mostly copying how @zou3519 did it in Python torch dispatch mode. To directly copy SafePyObject need to teach PyInterpreter how to incref so we can give SafePyObject copy constructor/assignment. I don't think it matters too much; less GIL acquisitions with the shared ptr here.

albanD replied:
Sounds good!

albanD reviewed (approved):
Only the remaining nits.
Logic sounds good otherwise, exciting to see more python-first implementations!

zou3519 reviewed (commented):
still making my way through, here are some initial comments

zou3519 commented on torch/overrides.py:
```diff
@@ -1587,3 +1601,229 @@ def is_tensor_like(inp):
+            a :class:`TorchFunctionMode`.  If your :class:`TorchFunctionMode`
+            has no ``__init__`` implementation, you can simply pass the class
+            itself (e.g., ``push_torch_function_mode(MyMode)``); otherwise,
+            use ``partial`` to partially apply the constructor with all
```
nit: partial -> functools.partial

To check my understanding, this is the approach to create a Mode that has some additional metadata attached?

ezyang replied:
Yup

zou3519 reviewed (approved):
LGTM. Had some minor questions and nits. Implementing most of the logic in Python makes things a lot simpler!

ezyang commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -183,11 +186,19 @@ auto combine_self_args(PyObject *self, PyObject *args) -> py::tuple {
   return args_;
 }
 
+// TODO: I'm not sure if I should call this __torch_function__ or
+// torch_function.  The former makes it easier to take an existing
+// Tensor-like __torch_function__ object and turn it into a mode;
+// but in general modes don't have to be Tensor-like (and we will
+// improperly accept mode objects as arguments when they shouldn't
```
Yes, if we change the name then we need a special case for "pass Tensor subclass as a mode" so that we look at the old name. Score another point for the "reuse the same name" team

ezyang committed:
code review on "__torch_function__ mode"


`__torch_function__` mode brings `enable_python_mode` functionality to the `__torch_function__` level of overrides (at the Python API level). I intend to use this to collect test data for validating abstract interpreters we plan to write for operators, but there are many other possible applications (for example, tracers like FX could use this to track factory function calls without monkey-patching the PyTorch namespace, and the mode could be used to vary backward semantics of functions without having to add new arguments to them, see https://github.com/albanD/subclass_zoo/blob/main/inner_autograd_tensor.py)

This PR also gives a design for mode stacks which is implemented in Python (rather in some sort of C++ infrastructure). My goal here was to trade off performance so that I could simplify the C++ bindings (we only expose `_set_torch_function_mode` and `_get_torch_function_mode` in the bindings) and iterate quickly on a more sophisticated mode stack design. The design here is oriented around TorchFunctionMode objects, which are chained together via object composition. IF this design is accepted, we should also apply it to the existing `__torch_dispatch__` mode.

**What is `__torch_function__` mode?** Check the documentation for TorchFunctionMode in torch/overrides.py

**How does it work?** There are two parts to the implementation, a simple but non-compositional C++ API, and a compositional Python API built on top of the C++ API.

In C++:

* We add a new Python mode object to PythonTorchFunctionTLS; the object doesn't have any type constraints but we will assume that it has a `__torch_function__` method defined on it. (Unlike `python_eager_mode`, this object need not be a class; it can be an honest to goodness instance of a class.)
* We modify `has_torch_function` to return true if the mode is non-null. In some cases, this is wrong (specifically when we are accumulating overloaded arguments), so those get a special argument to say "ignore modes when testing." I didn't do a very careful audit of all the sites so this is probably where we mostly likely have bugs in the implementation.
* When handling `__torch_function__`, we check if this mode is set. If it is, we first call `__torch_function__` on the mode to attempt to handle the operation (that is to say, we process modes first, and then concrete values)--unusually, we also reset the mode to be empty before calling back into `__torch_function__` (see https://github.com/pytorch/pytorch/issues/75166 for back story). If the mode returns NotImplemented we attempt to handle the operator with the subclasses instead. I fixed a bug in the error reporting where pointer values rather than actual reprs were printed.

In Python:

* We add a new class hierarchy TorchFunctionMode, representing objects that can handle `__torch_function__` and can chain with each other (allowing for nested modes). These classes have a metaclass TorchFunctionModeMeta which is (1) responsible for adding chaining functionality, without requiring users to manually record `inner` themselves, and (2) responsible for resetting the current mode (which is null at the time `__torch_function__` is called) to the mode of the *inner* mode class. We represent `__torch_function__` handlers as objects, not classes, so that a single handler can be associated with arbitrary state (most notably, the `inner` property; but you can define a parametrized modes--see example in test suite).
* We then provide two user-visible functions for working with the mode stack: `push_torch_function_mode` is the recommended function which is compositional (allowing for nested modes) as well as `enable_torch_function_mode` which is non-compositional (good for simple use cases, or swapping out a known mode with another mode). Check the docblocks in the PR for more details.
* For symmetry with `__torch_dispatch__` mode, we also accept a Tensor subclass as a mode passed to the argument; this works in much the same way a non-compositional TorchFunctionMode would be applied.

**Where there alternatives?**

* `push_torch_function_mode` could have been implemented as a C++ level concept, representing the stack explicitly in C++. This would have involved writing quite a lot of complex C++ binding code, and when I determined that I could implement all of this in Python using only two simple bindings I opted for this strategy for development speed. It may be useful to do it this way in the future for performance reasons.
* This PR introduces a new concept, TorchFunctionMode, whereas the existing `__torch_dispatch__` mode reused the class object for specifying the mode. I think introducing a new concept here is warranted for the following reasons: (1) `enable_python_mode` awkwardly requires you to subclass Tensor to write your mode, even if you never intend to instantiate your tensor, (2) there is no way to add extra metadata to a class without dynamically generating a class on the fly; this metadata is required for compositional use cases (aka, `inner` ala functorch), (3) even if you were able to dynamically generate the class working up and sideways the class hierarchy with super() and CurrentClass is awkward and people are unlikely to do it correctly. This is a case of favoring composition over inheritance, giving us an easier to understand API (even if it is not parsimonious).

**Examples?** Check the test cases in TestTorchFunctionMode in test/test_overrides.py

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "__torch_function__ mode"


`__torch_function__` mode brings `enable_python_mode` functionality to the `__torch_function__` level of overrides (at the Python API level). I intend to use this to collect test data for validating abstract interpreters we plan to write for operators, but there are many other possible applications (for example, tracers like FX could use this to track factory function calls without monkey-patching the PyTorch namespace, and the mode could be used to vary backward semantics of functions without having to add new arguments to them, see https://github.com/albanD/subclass_zoo/blob/main/inner_autograd_tensor.py)

This PR also gives a design for mode stacks which is implemented in Python (rather in some sort of C++ infrastructure). My goal here was to trade off performance so that I could simplify the C++ bindings (we only expose `_set_torch_function_mode` and `_get_torch_function_mode` in the bindings) and iterate quickly on a more sophisticated mode stack design. The design here is oriented around TorchFunctionMode objects, which are chained together via object composition. IF this design is accepted, we should also apply it to the existing `__torch_dispatch__` mode.

**What is `__torch_function__` mode?** Check the documentation for TorchFunctionMode in torch/overrides.py

**How does it work?** There are two parts to the implementation, a simple but non-compositional C++ API, and a compositional Python API built on top of the C++ API.

In C++:

* We add a new Python mode object to PythonTorchFunctionTLS; the object doesn't have any type constraints but we will assume that it has a `__torch_function__` method defined on it. (Unlike `python_eager_mode`, this object need not be a class; it can be an honest to goodness instance of a class.)
* We modify `has_torch_function` to return true if the mode is non-null. In some cases, this is wrong (specifically when we are accumulating overloaded arguments), so those get a special argument to say "ignore modes when testing." I didn't do a very careful audit of all the sites so this is probably where we mostly likely have bugs in the implementation.
* When handling `__torch_function__`, we check if this mode is set. If it is, we first call `__torch_function__` on the mode to attempt to handle the operation (that is to say, we process modes first, and then concrete values)--unusually, we also reset the mode to be empty before calling back into `__torch_function__` (see https://github.com/pytorch/pytorch/issues/75166 for back story). If the mode returns NotImplemented we attempt to handle the operator with the subclasses instead. I fixed a bug in the error reporting where pointer values rather than actual reprs were printed.

In Python:

* We add a new class hierarchy TorchFunctionMode, representing objects that can handle `__torch_function__` and can chain with each other (allowing for nested modes). These classes have a metaclass TorchFunctionModeMeta which is (1) responsible for adding chaining functionality, without requiring users to manually record `inner` themselves, and (2) responsible for resetting the current mode (which is null at the time `__torch_function__` is called) to the mode of the *inner* mode class. We represent `__torch_function__` handlers as objects, not classes, so that a single handler can be associated with arbitrary state (most notably, the `inner` property; but you can define a parametrized modes--see example in test suite).
* We then provide two user-visible functions for working with the mode stack: `push_torch_function_mode` is the recommended function which is compositional (allowing for nested modes) as well as `enable_torch_function_mode` which is non-compositional (good for simple use cases, or swapping out a known mode with another mode). Check the docblocks in the PR for more details.
* For symmetry with `__torch_dispatch__` mode, we also accept a Tensor subclass as a mode passed to the argument; this works in much the same way a non-compositional TorchFunctionMode would be applied.

**Where there alternatives?**

* `push_torch_function_mode` could have been implemented as a C++ level concept, representing the stack explicitly in C++. This would have involved writing quite a lot of complex C++ binding code, and when I determined that I could implement all of this in Python using only two simple bindings I opted for this strategy for development speed. It may be useful to do it this way in the future for performance reasons.
* This PR introduces a new concept, TorchFunctionMode, whereas the existing `__torch_dispatch__` mode reused the class object for specifying the mode. I think introducing a new concept here is warranted for the following reasons: (1) `enable_python_mode` awkwardly requires you to subclass Tensor to write your mode, even if you never intend to instantiate your tensor, (2) there is no way to add extra metadata to a class without dynamically generating a class on the fly; this metadata is required for compositional use cases (aka, `inner` ala functorch), (3) even if you were able to dynamically generate the class working up and sideways the class hierarchy with super() and CurrentClass is awkward and people are unlikely to do it correctly. This is a case of favoring composition over inheritance, giving us an easier to understand API (even if it is not parsimonious).

**Examples?** Check the test cases in TestTorchFunctionMode in test/test_overrides.py

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

peterbell10 commented on torch/csrc/utils/python_arg_parser.cpp:
```diff
@@ -278,6 +278,12 @@ auto handle_torch_function_no_python_arg_parser(
   if (ret.ptr() == nullptr || ret.ptr() == Py_NotImplemented) {
     for (auto& arg : overloaded_args) {
+      // See https://github.com/pytorch/pytorch/issues/63767
+      if (PyObject_FastGetAttrString(torch_function.ptr(), "__self__").is(arg)) {
```
This uses `torch_function` so needs to be below where that variable is set.

ezyang committed:
Update on "__torch_function__ mode"


`__torch_function__` mode brings `enable_python_mode` functionality to the `__torch_function__` level of overrides (at the Python API level). I intend to use this to collect test data for validating abstract interpreters we plan to write for operators, but there are many other possible applications (for example, tracers like FX could use this to track factory function calls without monkey-patching the PyTorch namespace, and the mode could be used to vary backward semantics of functions without having to add new arguments to them, see https://github.com/albanD/subclass_zoo/blob/main/inner_autograd_tensor.py)

This PR also gives a design for mode stacks which is implemented in Python (rather in some sort of C++ infrastructure). My goal here was to trade off performance so that I could simplify the C++ bindings (we only expose `_set_torch_function_mode` and `_get_torch_function_mode` in the bindings) and iterate quickly on a more sophisticated mode stack design. The design here is oriented around TorchFunctionMode objects, which are chained together via object composition. IF this design is accepted, we should also apply it to the existing `__torch_dispatch__` mode.

**What is `__torch_function__` mode?** Check the documentation for TorchFunctionMode in torch/overrides.py

**How does it work?** There are two parts to the implementation, a simple but non-compositional C++ API, and a compositional Python API built on top of the C++ API.

In C++:

* We add a new Python mode object to PythonTorchFunctionTLS; the object doesn't have any type constraints but we will assume that it has a `__torch_function__` method defined on it. (Unlike `python_eager_mode`, this object need not be a class; it can be an honest to goodness instance of a class.)
* We modify `has_torch_function` to return true if the mode is non-null. In some cases, this is wrong (specifically when we are accumulating overloaded arguments), so those get a special argument to say "ignore modes when testing." I didn't do a very careful audit of all the sites so this is probably where we mostly likely have bugs in the implementation.
* When handling `__torch_function__`, we check if this mode is set. If it is, we first call `__torch_function__` on the mode to attempt to handle the operation (that is to say, we process modes first, and then concrete values)--unusually, we also reset the mode to be empty before calling back into `__torch_function__` (see https://github.com/pytorch/pytorch/issues/75166 for back story). If the mode returns NotImplemented we attempt to handle the operator with the subclasses instead. I fixed a bug in the error reporting where pointer values rather than actual reprs were printed.

In Python:

* We add a new class hierarchy TorchFunctionMode, representing objects that can handle `__torch_function__` and can chain with each other (allowing for nested modes). These classes have a metaclass TorchFunctionModeMeta which is (1) responsible for adding chaining functionality, without requiring users to manually record `inner` themselves, and (2) responsible for resetting the current mode (which is null at the time `__torch_function__` is called) to the mode of the *inner* mode class. We represent `__torch_function__` handlers as objects, not classes, so that a single handler can be associated with arbitrary state (most notably, the `inner` property; but you can define a parametrized modes--see example in test suite).
* We then provide two user-visible functions for working with the mode stack: `push_torch_function_mode` is the recommended function which is compositional (allowing for nested modes) as well as `enable_torch_function_mode` which is non-compositional (good for simple use cases, or swapping out a known mode with another mode). Check the docblocks in the PR for more details.
* For symmetry with `__torch_dispatch__` mode, we also accept a Tensor subclass as a mode passed to the argument; this works in much the same way a non-compositional TorchFunctionMode would be applied.

**Where there alternatives?**

* `push_torch_function_mode` could have been implemented as a C++ level concept, representing the stack explicitly in C++. This would have involved writing quite a lot of complex C++ binding code, and when I determined that I could implement all of this in Python using only two simple bindings I opted for this strategy for development speed. It may be useful to do it this way in the future for performance reasons.
* This PR introduces a new concept, TorchFunctionMode, whereas the existing `__torch_dispatch__` mode reused the class object for specifying the mode. I think introducing a new concept here is warranted for the following reasons: (1) `enable_python_mode` awkwardly requires you to subclass Tensor to write your mode, even if you never intend to instantiate your tensor, (2) there is no way to add extra metadata to a class without dynamically generating a class on the fly; this metadata is required for compositional use cases (aka, `inner` ala functorch), (3) even if you were able to dynamically generate the class working up and sideways the class hierarchy with super() and CurrentClass is awkward and people are unlikely to do it correctly. This is a case of favoring composition over inheritance, giving us an easier to understand API (even if it is not parsimonious).

**Examples?** Check the test cases in TestTorchFunctionMode in test/test_overrides.py

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge this
